File: app/src/main/java/com/jgm90/cloudmusic/activities/BaseActivity.java
package com.jgm90.cloudmusic.activities;

import android.app.Service;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.media.AudioManager;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.jgm90.cloudmusic.R;
import com.jgm90.cloudmusic.fragments.PlaybackControlsFragment;
import com.jgm90.cloudmusic.services.MediaPlayerService;
import com.jgm90.cloudmusic.utils.DbHelper;

public class BaseActivity extends AppCompatActivity implements ServiceConnection {

    protected MediaPlayerService player_service;
    protected boolean serviceBound = false;

    private PlaybackControlsFragment mControlsFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setVolumeControlStream(AudioManager.STREAM_MUSIC);
        new DbHelper(this);
        attachService();
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.d("Base", "OnStart");
        mControlsFragment = (PlaybackControlsFragment) getFragmentManager()
                .findFragmentById(R.id.fragment_playback_controls);
        if (mControlsFragment == null) {
            //throw new IllegalStateException("Mising fragment with id 'controls'. Cannot continue.");
        } else {
            if (player_service == null) {
                hidePlaybackControls();
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (player_service != null && player_service.current_index() != -1) {
            showPlaybackControls();
        }
    }

    protected void showPlaybackControls() {
        Log.d("Base", "showPlaybackControls");
        mControlsFragment = (PlaybackControlsFragment) getFragmentManager()
                .findFragmentById(R.id.fragment_playback_controls);
        if (mControlsFragment != null) {
            getFragmentManager().beginTransaction()
                    .show(mControlsFragment)
                    .commit();
        }
    }

    protected void hidePlaybackControls() {
        Log.d("Base", "hidePlaybackControls");
        getFragmentManager().beginTransaction()
                .hide(mControlsFragment)
                .commit();
    }

    private void attachService() {
        Intent service = new Intent(this, MediaPlayerService.class);
        bindService(service, this, Service.BIND_AUTO_CREATE);
    }

    private void detachService() {
        unbindService(this);
    }

    protected void onServiceAttached(MediaPlayerService service) {
        // do something necessary by its subclass.
        if (service.isPlaying()) {
            showPlaybackControls();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onServiceConnected(ComponentName name, IBinder service) {
        MediaPlayerService.LocalBinder serviceBinder = (MediaPlayerService.LocalBinder) service;
        player_service = serviceBinder.getService();
        serviceBound = true;
        onServiceAttached(player_service);
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {
        serviceBound = false;
        player_service = null;
    }

    @Override
    protected void onDestroy() {
        detachService();
        super.onDestroy();
    }
}

File: app/src/main/java/com/jgm90/cloudmusic/activities/BaseActivity.kt
package com.jgm90.cloudmusic.activities

import android.app.Service
import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.media.AudioManager
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.jgm90.cloudmusic.R
import com.jgm90.cloudmusic.fragments.PlaybackControlsFragment
import com.jgm90.cloudmusic.services.MediaPlayerService
import com.jgm90.cloudmusic.utils.DbHelper

open class BaseActivity : AppCompatActivity(), ServiceConnection {
    protected var player_service: MediaPlayerService? = null
    protected var serviceBound = false

    private var controlsFragment: PlaybackControlsFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        volumeControlStream = AudioManager.STREAM_MUSIC
        DbHelper.getInstance(this)
        attachService()
    }

    override fun onStart() {
        super.onStart()
        Log.d("Base", "OnStart")
        controlsFragment = fragmentManager.findFragmentById(R.id.fragment_playback_controls) as? PlaybackControlsFragment
        if (controlsFragment != null) {
            if (player_service == null) {
                hidePlaybackControls()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        if (player_service != null && player_service?.current_index() != -1) {
            showPlaybackControls()
        }
    }

    protected fun showPlaybackControls() {
        Log.d("Base", "showPlaybackControls")
        controlsFragment = fragmentManager.findFragmentById(R.id.fragment_playback_controls) as? PlaybackControlsFragment
        controlsFragment?.let {
            fragmentManager.beginTransaction().show(it).commit()
        }
    }

    protected fun hidePlaybackControls() {
        Log.d("Base", "hidePlaybackControls")
        controlsFragment?.let {
            fragmentManager.beginTransaction().hide(it).commit()
        }
    }

    private fun attachService() {
        val service = Intent(this, MediaPlayerService::class.java)
        bindService(service, this, Service.BIND_AUTO_CREATE)
    }

    private fun detachService() {
        unbindService(this)
    }

    protected open fun onServiceAttached(service: MediaPlayerService) {
        if (service.isPlaying()) {
            showPlaybackControls()
        }
    }

    override fun onServiceConnected(name: ComponentName, service: IBinder) {
        val serviceBinder = service as MediaPlayerService.LocalBinder
        player_service = serviceBinder.getService()
        serviceBound = true
        player_service?.let { onServiceAttached(it) }
    }

    override fun onServiceDisconnected(name: ComponentName) {
        serviceBound = false
        player_service = null
    }

    override fun onDestroy() {
        detachService()
        super.onDestroy()
    }
}
